File: harness/java/src/main/java/com/azure/cosmos/testrunner/Assertions.java
package com.azure.cosmos.testrunner;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Evaluates step assertions; mirrors the Python assertions module. */
public final class Assertions {

    private Assertions() {
    }

    @SuppressWarnings("unchecked")
    public static List<Map<String, Object>> evaluate(List<Map<String, Object>> expectations,
                                                     OpResult result, Backend backend,
                                                     Map<String, Object> context) {
        List<Map<String, Object>> outcomes = new ArrayList<>();
        if (expectations == null) return outcomes;
        for (Map<String, Object> exp : expectations) {
            outcomes.add(evaluateOne(exp, result, backend, context));
        }
        return outcomes;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> evaluateOne(Map<String, Object> exp, OpResult result,
                                                   Backend backend, Map<String, Object> context) {
        String type = String.valueOf(exp.get("type"));
        Object nameObj = exp.getOrDefault("name", type);
        String name = String.valueOf(nameObj);

        switch (type) {
            case "ok":
            case "no_error":
                return outcome(name, result.ok, result.ok ? "" : "status=" + result.statusCode);
            case "error":
                return outcome(name, !result.ok, result.ok ? "expected failure but op succeeded" : "");
            case "status_code":
                return outcome(name, result.statusCode == asInt(exp.get("value")), "actual=" + result.statusCode);
            case "error_status":
                return outcome(name, !result.ok && result.statusCode == asInt(exp.get("value")),
                        "ok=" + result.ok + " status=" + result.statusCode);
            case "item_count": {
                int actual = result.items == null ? 0 : result.items.size();
                return outcome(name, actual == asInt(exp.get("value")), "actual=" + actual);
            }
            case "count_gte": {
                int actual = result.items == null ? 0 : result.items.size();
                return outcome(name, actual >= asInt(exp.get("value")), "actual=" + actual);
            }
            case "field_equals": {
                Object actual = navigate(result, String.valueOf(exp.get("path")));
                return outcome(name, String.valueOf(actual).equals(String.valueOf(exp.get("value"))),
                        "actual=" + actual);
            }
            case "metric_equals": {
                Object actual = metric(backend, String.valueOf(exp.get("metric")));
                return outcome(name, String.valueOf(actual).equals(String.valueOf(exp.get("value"))),
                        "actual=" + actual);
            }
            case "metric_zero": {
                Object actual = metric(backend, String.valueOf(exp.get("metric")));
                return outcome(name, asInt(actual) == 0, "actual=" + actual);
            }

            // --- transport / protocol fault assertions (emulator/live tier) ---- //
            case "eventually_ok":
                // Op succeeded (after any injected fault was cleared / retried).
                return outcome(name, result.ok, result.ok ? "" : "status=" + result.statusCode);
            case "retry_count_gte": {
                int actual = asInt(metric(backend, "retries"));
                return outcome(name, actual >= asInt(exp.get("value")), "retries=" + actual);
            }
            case "latency_percentile": {
                String scope = scope(exp, context);
                List<Double> samples = samples(context, "latency", scope);
                if (samples.isEmpty()) {
                    return outcome(name, false, "no latency samples for scope '" + scope + "'");
                }
                int pct = exp.get("percentile") != null ? asInt(exp.get("percentile")) : 99;
                double p = percentile(samples, pct);
                return outcome(name, p <= asDouble(exp.get("max_ms")),
                        String.format("p%d=%.1fms n=%d", pct, p, samples.size()));
            }
            case "resource_stable": {
                String scope = scope(exp, context);
                List<Double> samples = samples(context, "resource", scope);
                if (samples.size() < 2) {
                    return outcome(name, false, "insufficient resource samples for scope '" + scope + "'");
                }
                double lo = samples.stream().mapToDouble(Double::doubleValue).min().orElse(0);
                double hi = samples.stream().mapToDouble(Double::doubleValue).max().orElse(0);
                double drift = hi == 0 ? 0.0 : (hi - lo) / hi * 100.0;
                double tol = exp.get("tolerance_pct") != null ? asDouble(exp.get("tolerance_pct")) : 10.0;
                return outcome(name, drift <= tol,
                        String.format("drift=%.1f%% (min=%s max=%s n=%d)", drift, lo, hi, samples.size()));
            }
            case "failover_region": {
                // Best-effort: the contacted region is surfaced in SDK diagnostics on
                // emulator/live. Offline/mock has no diagnostics, so this fails clearly.
                String want = String.valueOf(exp.getOrDefault("equals", ""));
                String blob = result.diagnostics != null ? result.diagnostics : "";
                boolean present = !want.isEmpty() && blob.contains(want);
                return outcome(name, present, "want_region='" + want + "' present=" + present);
            }
            case "pages_cover_set": {
                List<String> ids = new ArrayList<>();
                if (result.items != null) {
                    for (Object r : result.items) {
                        if (r instanceof Map) {
                            ids.add(String.valueOf(((Map<String, Object>) r).get("id")));
                        }
                    }
                }
                java.util.Set<String> unique = new java.util.HashSet<>(ids);
                boolean noDupes = ids.size() == unique.size();
                Object expected = exp.get("expected_count");
                boolean covers = expected == null || unique.size() == asInt(expected);
                return outcome(name, noDupes && covers,
                        "n=" + ids.size() + " unique=" + unique.size() + " expected=" + expected
                                + " no_dupes=" + noDupes);
            }

            default:
                return outcome(name, false, "unknown assertion type '" + type + "'");
        }
    }

    @SuppressWarnings("unchecked")
    private static String scope(Map<String, Object> exp, Map<String, Object> context) {
        Object s = exp.get("scope");
        if (s != null) return String.valueOf(s);
        return context != null && context.get("scope") != null ? String.valueOf(context.get("scope")) : null;
    }

    @SuppressWarnings("unchecked")
    private static List<Double> samples(Map<String, Object> context, String key, String scope) {
        if (context == null) return new ArrayList<>();
        Object byScope = context.get(key);
        if (!(byScope instanceof Map)) return new ArrayList<>();
        Object list = ((Map<String, Object>) byScope).get(scope);
        return list instanceof List ? (List<Double>) list : new ArrayList<>();
    }

    private static double percentile(List<Double> samples, int pct) {
        List<Double> sorted = new ArrayList<>(samples);
        java.util.Collections.sort(sorted);
        if (sorted.size() == 1) return sorted.get(0);
        double rank = pct / 100.0 * (sorted.size() - 1);
        int lo = (int) Math.floor(rank);
        int hi = (int) Math.ceil(rank);
        double frac = rank - lo;
        return sorted.get(lo) + (sorted.get(hi) - sorted.get(lo)) * frac;
    }

    private static double asDouble(Object v) {
        if (v instanceof Number) return ((Number) v).doubleValue();
        return Double.parseDouble(String.valueOf(v));
    }

    @SuppressWarnings("unchecked")
    private static Object navigate(OpResult result, String path) {
        Object cur = result.asContext();
        for (String part : path.split("\\.")) {
            if (cur == null) return null;
            if (cur instanceof List) {
                cur = ((List<Object>) cur).get(Integer.parseInt(part));
            } else if (cur instanceof Map) {
                cur = ((Map<String, Object>) cur).get(part);
            } else {
                return null;
            }
        }
        return cur;
    }

    @SuppressWarnings("unchecked")
    private static Object metric(Backend backend, String name) {
        Object cur = backend.metrics().asMap();
        for (String part : name.split("\\.")) {
            if (cur instanceof Map) {
                cur = ((Map<String, Object>) cur).get(part);
            } else {
                return null;
            }
        }
        return cur;
    }

    private static int asInt(Object v) {
        if (v instanceof Number) return ((Number) v).intValue();
        return Integer.parseInt(String.valueOf(v));
    }

    private static Map<String, Object> outcome(String name, boolean passed, String detail) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("name", name);
        m.put("passed", passed);
        m.put("detail", detail);
        return m;
    }
}

File: harness/python/cosmos_test_runner/assertions.py
"""Assertion evaluation for scenario steps."""

from __future__ import annotations

from typing import Any, Dict, List

from .backends import Backend, OpResult


def _navigate(result: OpResult, path: str) -> Any:
    """Resolve a dotted path against an OpResult (e.g. 'item.total', 'items.0.id')."""
    root: Dict[str, Any] = {
        "item": result.item,
        "items": result.items,
        "status_code": result.status_code,
        "error_code": result.error_code,
        "ok": result.ok,
    }
    cur: Any = root
    for part in path.split("."):
        if cur is None:
            return None
        if isinstance(cur, list):
            cur = cur[int(part)]
        elif isinstance(cur, dict):
            cur = cur.get(part)
        else:
            return None
    return cur


def _metric(backend: Backend, name: str) -> Any:
    cur: Any = backend.metrics.as_dict()
    for part in name.split("."):
        if isinstance(cur, dict):
            cur = cur.get(part)
        else:
            return None
    return cur


def evaluate(expectations: List[Dict[str, Any]], result: OpResult, backend: Backend,
             snapshots: Dict[str, Dict[str, Any]] = None) -> List[Dict[str, Any]]:
    """Return a list of assertion outcomes ({name, passed, detail}).

    ``snapshots`` maps step id -> metrics dict captured after that step ran; it
    powers span assertions like ``metric_delta`` that compare two points in time.
    """
    outcomes: List[Dict[str, Any]] = []
    for exp in expectations or []:
        outcomes.append(_evaluate_one(exp, result, backend, snapshots or {}))
    return outcomes


def _metric_from(snapshot: Dict[str, Any], path: str) -> Any:
    cur: Any = snapshot
    for part in path.split("."):
        if isinstance(cur, dict):
            cur = cur.get(part)
        else:
            return None
    return cur


def _evaluate_one(exp: Dict[str, Any], result: OpResult, backend: Backend,
                  snapshots: Dict[str, Dict[str, Any]]) -> Dict[str, Any]:
    t = exp.get("type")
    name = exp.get("name") or t

    def outcome(passed: bool, detail: str = "") -> Dict[str, Any]:
        return {"name": name, "passed": bool(passed), "detail": detail}

    if t in ("ok", "no_error"):
        return outcome(result.ok, f"status={result.status_code}" if not result.ok else "")
    if t == "error":
        return outcome(not result.ok, "expected failure but op succeeded" if result.ok else f"got {result.status_code}")
    if t == "status_code":
        return outcome(result.status_code == exp["value"], f"actual={result.status_code}")
    if t == "error_status":
        return outcome((not result.ok) and result.status_code == exp["value"],
                       f"ok={result.ok} status={result.status_code}")
    if t == "item_count":
        actual = len(result.items or [])
        return outcome(actual == exp["value"], f"actual={actual}")
    if t == "count_gte":
        actual = len(result.items or [])
        return outcome(actual >= exp["value"], f"actual={actual}")
    if t == "field_equals":
        actual = _navigate(result, exp["path"])
        return outcome(str(actual) == str(exp["value"]), f"actual={actual!r}")
    if t == "metric_equals":
        actual = _metric(backend, exp["name_path"] if "name_path" in exp else exp["metric"])
        return outcome(actual == exp["value"], f"actual={actual!r}")
    if t == "metric_zero":
        actual = _metric(backend, exp["metric"])
        return outcome(actual == 0, f"actual={actual!r}")

    # --- control-plane assertions ---------------------------------------- #
    if t == "metric_delta":
        # Compare a metric between two step snapshots: value == snap[b] - snap[a].
        over = exp.get("over") or []
        if len(over) != 2:
            return outcome(False, "metric_delta requires over: [stepA, stepB]")
        step_a, step_b = over
        snap_a, snap_b = snapshots.get(step_a), snapshots.get(step_b)
        if snap_a is None or snap_b is None:
            return outcome(False, f"missing snapshot(s) for {over} (have {list(snapshots)})")
        a = _metric_from(snap_a, exp["metric"]) or 0
        b = _metric_from(snap_b, exp["metric"]) or 0
        delta = b - a
        return outcome(delta == exp["value"], f"delta={delta} (a={a} b={b})")
    if t == "sequence":
        of = exp.get("of", "status_codes")
        actual = list(result.status_sequence if of == "status_codes" else result.metadata_events)
        if "equals" in exp:
            want = exp["equals"]
            return outcome(actual == want, f"actual={actual} want={want}")
        if "contains_in_order" in exp:
            want = exp["contains_in_order"]
            ok = _contains_in_order(actual, want)
            return outcome(ok, f"actual={actual} expected_subsequence={want}")
        return outcome(False, "sequence needs 'equals' or 'contains_in_order'")
    if t == "page_size_at_most":
        actual = len(result.items or [])
        return outcome(actual <= exp["value"], f"actual={actual}")

    return outcome(False, f"unknown assertion type '{t}'")


def _contains_in_order(actual: List[Any], want: List[Any]) -> bool:
    """True if ``want`` appears as an ordered (not necessarily contiguous) subsequence."""
    it = iter(actual)
    return all(any(x == w for x in it) for w in want)
